File: src/AirFunctions/Admin/AdminFlights.java
package AirFunctions.Admin;

import AirFunctions.AccountFunctions;
import AirFunctions.AirplaneFunctions;
import AirFunctions.Calculations;
import AirFunctions.CityFunctions;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.sql.Connection;
import java.util.ArrayList;


/**
 * Created by johnn on 4/8/2017.
 */
@WebServlet("/AirFunctions.Admin.AdminFlights")
public class AdminFlights extends HttpServlet {

    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException  {

        try {
            String departState = request.getParameter("flightDepartureLocationState");
            String departCity = request.getParameter("flightDepartureLocationCity");
            String arrivalState = request.getParameter("flightArrivalLocationState");
            String arrivalCity = request.getParameter("flightArrivalLocationCity");
            String planeModel = request.getParameter("flightPlaneModelSelect");

            System.out.println("-----------------------");
            System.out.println("departState: " + departState);
            System.out.println("departCity: " + departCity);
            System.out.println("arrivalState: " + arrivalState);
            System.out.println("arrivalCity: " + arrivalCity);
            System.out.println("planeModel: " + planeModel);

            System.out.println(request);
            System.out.println(request.getContextPath());
            System.out.println(request.getQueryString());

            double[] departLatAndLong = getLatAndLong(departState, departCity);
            double[] arrivalLatAndLong = getLatAndLong(arrivalState, arrivalCity);
            double distance = Calculations.getDistanceBetweenCitiesInKilometers(
                    departLatAndLong[0], departLatAndLong[1],
                    arrivalLatAndLong[0], arrivalLatAndLong[1]);

            String[] planeSpecifics = getPlaneSpecifics(planeModel);
            int passengerCapacity = Integer.parseInt(planeSpecifics[0]);
            int fuelCap = Integer.parseInt(planeSpecifics[1]);
            double fuelBurnRate = Double.parseDouble(planeSpecifics[2]);
            int averageVelocity = Integer.parseInt(planeSpecifics[3]);


            int canTravel = (Calculations.hasEnoughFuel(fuelCap, fuelBurnRate, distance)) ? 1: 0;
            int time = (int)(distance * 60 / averageVelocity);
            int distancePrice = Calculations.getDistancePrice(distance);

            String responseText = "{\"canTravel\":"+canTravel+",\"timed\":"+time+",\"distancePrice\":"+distancePrice+"}";

            System.out.println(responseText);

            response.setContentType("text/plain");
            response.getWriter().print(responseText);
            System.out.println("-----------------------");

        }catch(Exception e){
            e.printStackTrace();
        }
    }

    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        try {
            String departDate = request.getParameter("flightDepartureDate");
            String departTime = request.getParameter("flightDepartureTime");
            String arrivalDate = request.getParameter("flightArrivalDate");
            String arrivalTime = request.getParameter("flightArrivalTime");
            String planeModel = request.getParameter("flightPlaneModelSelect");

            System.out.println("-----------------------");
            System.out.println("departDate: " + departDate);
            System.out.println("departTime: " + departTime);
            System.out.println("arrivalDate: " + arrivalDate);
            System.out.println("arrivalTime: " + arrivalTime);
            System.out.println("planeModel: " + planeModel);
            System.out.println("-----------------------");

            String json = compileJSON(getAvailableAirPlanes(planeModel,departDate,
                    departTime,arrivalDate,arrivalTime));

            System.out.println(json);
            response.setContentType("text/plain");
            response.getWriter().print(json);

        }catch(Exception e){
            e.printStackTrace();
        }
    }


    private static String[] getPlaneSpecifics(String planeModel){
        Connection con = AccountFunctions.OpenDatabase();
        String[] value = AirplaneFunctions.planeModelSpecifics(con, planeModel);
        AccountFunctions.closeConnection(con);
        return value;
    }

    private static double[] getLatAndLong(String state, String city) {
        double[] value = {0.0, 0.0};
        Connection con = AccountFunctions.OpenDatabase();
        String[] valueString = CityFunctions.getLatAndLong(con, state, city);
        value[0] = Double.parseDouble(valueString[0]);
        value[1] = Double.parseDouble(valueString[1]);
        AccountFunctions.closeConnection(con);

        return value;
    }

    private static ArrayList<String[]> getAvailableAirPlanes(String planeModel, String departDate, String departTime,
                                                             String arriveDate, String arriveTime){

        Connection con = AccountFunctions.OpenDatabase();
        ArrayList<String[]> planes = AirplaneFunctions.airplanesCurrentlyAvailable(con, planeModel, departDate, departTime,
                arriveDate, arriveTime);
        AccountFunctions.closeConnection(con);
        return planes;
    }

    private static String compileJSON(ArrayList<String[]> planeInfo){
        String[] list = {"planeType","id","capacity","maxEcon","maxBus","maxFirst","basePrice"};
        String jsonPlanes = "";
        boolean hasPlane = false;
        int numberOfPlanes = 0;
        for (int i = 0; i < planeInfo.size(); i++) {
            if (planeInfo.get(i)[0]!=null) {
                numberOfPlanes++;
                hasPlane = true;
                jsonPlanes += "{";
                jsonPlanes += "\"" + list[0] + "\":\"" + planeInfo.get(i)[0] + "\"";
                for (int j = 1; j < planeInfo.get(i).length; j++) {
                    jsonPlanes += ",\"" + list[j] + "\":" + planeInfo.get(i)[j];
                }
                jsonPlanes += "},";
            }
        }
        String json = "{\"numberOfPlanes\":"+numberOfPlanes+",\"planes\":";
        if(hasPlane){
            json += "["+jsonPlanes.substring(0,jsonPlanes.length()-1)+"]";
        }else{
            json += 0;
        }
        json += "}";
        return json;
    }

}

File: web/js/adminFlights.js
/**
 * Created by johnn on 4/11/2017.
 */

var availablePlaneList = "";

/**
 * Date picker for Flight Departure
 */
$( function() { $( "#flightDeparturedatepicker" ).datepicker({
    beforeShow: function(input,inst){
        var today = new Date();
        today.setDate(today.getDate()+1);
        $(this).datepicker('option','minDate',today);
    }
});} );

/**
 * Date picker for Flight Arrival
 */
$( function() { $( "#flightArrivaldatepicker" ).datepicker({
    beforeShow: function(input,inst){
        var today = new Date();
        today.setDate(today.getDate()+1);
        $(this).datepicker('option','minDate',today);
    }
});} );

/**
 * Time picker for Flight Departure
 */
$( function() {
    $( "#flightDeparturetimepicker" ).timepicker({
    timeFormat: 'HH:mm',
    minTime: new Date(0,0,0,0,0)
    });
} );

/**
 * Time picker for Flight Arrival
 */
$( function() {
    $( "#flightArrivaltimepicker" ).timepicker({
        timeFormat: 'HH:mm',
        minTime: new Date(0,0,0,0,0)
    });
} );

function enableNextEntry(number){
    if(number == 1){
        document.getElementById("flightDeparturetimepicker").disabled = false;
    } else if (number == 2){
        document.getElementById("flightDepartureLocationState").disabled = false;
    } else if (number == 3){
        document.getElementById("flightDepartureLocationCity").disabled = false;
    } else if (number == 4){
        document.getElementById("flightArrivalLocationState").disabled = false;
    } else if (number == 5){
        document.getElementById("flightArrivalLocationCity").disabled = false;
    } else if (number == 6){
        document.getElementById("flightPlaneModelSelect").disabled = false;
    } else if (number == 7){
        document.getElementById("addFlightButton").disabled = false;
    } else {
        document.getElementById("flightDeparturetimepicker").disabled = true;
        document.getElementById("flightDepartureLocationState").disabled = true;
        document.getElementById("flightDepartureLocationCity").disabled = true;
        document.getElementById("flightArrivalLocationState").disabled = true;
        document.getElementById("flightArrivalLocationCity").disabled = true;
        document.getElementById("flightPlaneModelSelect").disabled = true;
        document.getElementById("addFlightButton").disabled = true;
    }
}

/**
 *
 */
$( function() {
    $(".adminAddFlightModalButtons").onclick=function(){
        var prev = document.getElementById("adminAddFlightModalPreviousPage");
        var next = document.getElementById("adminAddFlightModalCurrentPage");
        if(this == prev){
            getAvailableAirplanes(prev);
        }else if(this == next){
            getAvailableAirplanes(next);
        }
    };
});

/**
 * Implemented to only show cities that go with the Departure State
 * @param classID or the class state ID
 */
function alterDepartureCitySelect(classID){
    var allCities = document.getElementById("flightDepartureLocationCity").getElementsByClassName("adminAllCities");

    for(var city=0; city<allCities.length; city++){
        allCities[city].style.display = "none";
    }

    var stateCities = document.getElementById("flightDepartureLocationCity").getElementsByClassName(classID);

    for(var stateCity=0; stateCity<stateCities.length; stateCity++){
        stateCities[stateCity].style.display = "block";
    }

}

/**
 * Implemented to only show cities that go with the Arrival State
 * @param classID or state ID class name
 */
function alterArrivalCitySelect(classID){

    var departStateSelect = document.getElementById("flightDepartureLocationState");
    var departState = departStateSelect.options[departStateSelect.selectedIndex].value;
    var departCitySelect = document.getElementById("flightDepartureLocationCity");
    var departCity = departCitySelect.options[departCitySelect.selectedIndex].value;
    var arrivalStateSelect = document.getElementById("flightArrivalLocationState");
    var arrivalState = arrivalStateSelect.options[arrivalStateSelect.selectedIndex].value;
    var allCities = document.getElementById("flightArrivalLocationCity").getElementsByClassName("adminAllCities");

    for(var city=0; city<allCities.length; city++){
        allCities[city].style.display = "none";
    }

    var stateCities = document.getElementById("flightArrivalLocationCity").getElementsByClassName(classID);

    for(var stateCity=0; stateCity<stateCities.length; stateCity++){
        stateCities[stateCity].style.display = "block";
        if(departState==arrivalState && departCity==stateCities[stateCity].value){
            stateCities[stateCity].style.display = "none";
        }
    }

}

/**
 * Gets the departure and arrival location information
 * @returns {*|jQuery} serialized information for jquery send
 */
function getLocationInfo(){
    var locations;
    locations =  $("#flightDepartureLocationState").serialize()+"&";
    locations +=  $("#flightDepartureLocationCity").serialize()+"&";
    locations +=  $("#flightArrivalLocationState").serialize()+"&";
    locations +=  $("#flightArrivalLocationCity").serialize();
    return locations;
}


/**
 * Determines if the selected plane model can travel the distance
 * "doGets" the information to "AirFunctions.Admin.AdminFlights"
 * The returned information includes a simple boolean, trip time,
 * and price adjustment due to distance traveled
 */
function canModelMakeTheDistance(){
    document.getElementById("flightDemandSlider").value = 1; // resets slider
    updateSliderText(); // updates slider text to new value
    availablePlaneList = "";
    var planeModelSelect = document.getElementById("flightPlaneModelSelect");
    var planeModel = planeModelSelect.options[planeModelSelect.selectedIndex].value;
    var series = getLocationInfo() + "&flightPlaneModelSelect="+planeModel;

   $.get(
        "AirFunctions.Admin.AdminFlights", series,
       function(msg){
           var ableTimedPrice = JSON.parse(msg);
            if(ableTimedPrice.canTravel == 1){
                modifyArrivalDateAndTime(ableTimedPrice.timed);
                document.getElementById("flightDistancePrice").value = ableTimedPrice.distancePrice;
                document.getElementById("adminAddFlightModalCurrentPage").innerHTML = "0";
                getAvailableAirplanes(0);
            } else{
                alert("this plane can't make the trip");
            }
       }
    );

}

/**
 * Sets the Arrival Date and Time once locations known.
 * @param tripTime
 */
function modifyArrivalDateAndTime(tripTime){
    var departDate = $("#flightDeparturedatepicker").datepicker('getDate');
    var departTime = $("#flightDeparturetimepicker").timepicker('getTime');
    var arrivalDate = new Date();

    console.log(tripTime);

    arrivalDate.setFullYear(departDate.getFullYear());
    arrivalDate.setMonth(departDate.getMonth());
    arrivalDate.setDate(departDate.getDate());
    arrivalDate.setHours(departTime.getHours());
    arrivalDate.setMinutes(departTime.getMinutes()+tripTime);

    $("#flightArrivaldatepicker").datepicker('setDate',arrivalDate);
    $("#flightArrivaltimepicker").timepicker('setTime',arrivalDate);
}


function getAvailableAirplanes(page){

    //String[] list = {"planeType","id","capacity","maxEcon","maxBus","maxFirst","basePrice"};
    $.post("AirFunctions.Admin.AdminFlights",getAvailabilityInfo(),
        function(msg){
            var response = JSON.parse(msg);
            removePreviousPlaneQuery();
            availablePlaneList = response.planes;
            if(response.numberOfPlanes == 0){
                // Code for when the number of planes that match the criteria is 0
            }else{
                setModalHeader();
                document.getElementById("adminAddFlightModalFormNumber").value = availablePlaneList.length;
                console.log(availablePlaneList.length);
                whatAirplanesToDisplay(0);
            }
        }
    )

}

function whatAirplanesToDisplay(page){


    for(var i=page*10; i<page*10+10; i++){
        if(i < availablePlaneList.length) {
            displayAvailablePlane(availablePlaneList[i].id);
        }
    }
    hideShowButtons(page)
}

/**
 * Gets the information to send to get list of available airplanes from server
 * @returns {string|*} of serialized information to pass to server
 */
function getAvailabilityInfo(){
    var departDate = $("#flightDeparturedatepicker").datepicker('getDate');
    var departTime = $("#flightDeparturetimepicker").timepicker('getTime');
    var arrivalDate = $("#flightArrivaldatepicker").datepicker('getDate');
    var arrivalTime = $("#flightArrivaltimepicker").timepicker('getTime');
    var planeModelSelect = document.getElementById("flightPlaneModelSelect");
    var nextSeries;
    nextSeries =    "flightDepartureDate="+returnDateInMSQLFormat(departDate,0) + "&";
    nextSeries +=   "flightDepartureTime="+returnTimeInMSQLFormat(departTime)+":00&";
    nextSeries +=   "flightArrivalDate="+returnDateInMSQLFormat(arrivalDate,0)+"&";
    nextSeries +=   "flightArrivalTime="+returnTimeInMSQLFormat(arrivalTime)+":00&";
    nextSeries +=   "flightPlaneModelSelect="+planeModelSelect.options[planeModelSelect.selectedIndex].text;
    return nextSeries;
}

function returnDateInMSQLFormat(date, adjust){
    var adjustMonth = date.getMonth() + adjust; // 0 if going to be parsed. 1 if going to be stored in database
    return date.getFullYear() + "-" + ((adjustMonth< 10) ? "0" : "") + adjustMonth + "-" + ((date.getDate() < 10) ? "0" : "") + date.getDate();
}

function returnTimeInMSQLFormat(time){
    return ((time.getHours() < 10) ? "0" : "") + time.getHours() + ":" + ((time.getMinutes() < 10) ? "0" : "") + time.getMinutes();
}

//String[] list = {"planeType","id","capacity","maxEcon","maxBus","maxFirst","basePrice"};
function displayAvailablePlane(id){
    var ul = document.getElementById("adminAddFlightModalUL");
    var li = document.createElement("li");
    var liDivPlane = document.createElement("div");
    var liDivCenter = document.createElement("div")
    var liDivResponse = document.createElement("div")
    var liDivP = document.createElement("p");
    var liDivInput = document.createElement("input");


    liDivInput.type="checkbox";
    liDivInput.class="addFlightCheckBoxes";
    liDivInput.value=id;
    liDivInput.onclick = function(){
        console.log(this.value);
      document.getElementById("adminAddFlightModalFormButton").value = this.value;
    };

    liDivP.innerHTML="ID: "+id;
    liDivCenter.innerHTML="";
    liDivPlane.appendChild(liDivP);
    liDivResponse.appendChild(liDivInput);

    li.appendChild(liDivPlane);
    li.appendChild(liDivCenter);
    li.appendChild(document.createElement("div"));
    li.appendChild(liDivResponse);
    li.name = "flightAddFlightFormOption";

    ul.appendChild(li);
}

/**
 * Opens up the modal that shows all available airplanes
 */
function viewAddFlightModal() {
    document.getElementById("adminAddFlightModal").style.display = "block";
}

/**
 * Closes the modal that shows all available airplanes
 */
function closeAddFlightModal(){
    document.getElementById("adminAddFlightModal").style.display = "none";
}

/**
 * Removes the previous searches available airplane list
 */
function removePreviousPlaneQuery(){
    var parent = document.getElementById("adminAddFlightModalUL");
    while(parent.hasChildNodes()) {
        parent.removeChild(parent.firstChild);
    }
}

function hideShowButtons(page){
    var current = document.getElementById("adminAddFlightModalCurrentPage");
    var prev = document.getElementById("adminAddFlightModalPreviousPage");
    var next = document.getElementById("adminAddFlightModalNextPage");
    var prevPage = page-1;
    current.innerHTML=page;
    if(prevPage<1){
        prev.hidden = true;
    }else{
        prev.show = true;
    }

    if(page == 0){
        current.hidden = true;
    } else {
        current.show = true;
    }

    if((page+1)*10<availablePlaneList.length){
        next.show = true;
    } else {
        next.hidden = true;
    }
}

function setModalHeader(){
    var planeModelSelect = document.getElementById("flightPlaneModelSelect");
    var departStateSelect = document.getElementById("flightDepartureLocationState");
    var departState = departStateSelect.options[departStateSelect.selectedIndex].value;
    var departCitySelect = document.getElementById("flightDepartureLocationCity");
    var departCity = departCitySelect.options[departCitySelect.selectedIndex].value;
    var arrivalStateSelect = document.getElementById("flightArrivalLocationState");
    var arrivalState = arrivalStateSelect.options[arrivalStateSelect.selectedIndex].value;
    var arrivalCitySelect = document.getElementById("flightArrivalLocationCity");
    var arrivalCity = arrivalCitySelect.options[arrivalCitySelect.selectedIndex].value;

    document.getElementById("adminFlightModalDepartDate").innerText = returnDateInMSQLFormat($("#flightDeparturedatepicker").datepicker('getDate'), 1);
    document.getElementById("adminFlightModalDepartTime").innerText = returnTimeInMSQLFormat($("#flightDeparturetimepicker").timepicker('getTime'));
    document.getElementById("adminFlightModalArriveDate").innerText = returnDateInMSQLFormat($("#flightArrivaldatepicker").datepicker('getDate'), 1);
    document.getElementById("adminFlightModalArriveTime").innerText = returnTimeInMSQLFormat($("#flightArrivaltimepicker").timepicker('getTime'));
    document.getElementById("adminFlightModalDepartLocation").innerText = departCity + ", " + departState;
    document.getElementById("adminFlightModalArriveLocation").innerText = arrivalCity + ", " + arrivalState;
    document.getElementById("adminAddFlightModalTitle").innerText = planeModelSelect.options[planeModelSelect.selectedIndex].text;
}


function submitToDatabase(){
    if(confirm("Are you sure?")){

        var data = getAvailabilityInfo() + "&" + getLocationInfo() + "&value=";
        data += document.getElementById("adminAddFlightModalFormButton").value;
        data += "&distancePrice="+document.getElementById("flightDistancePrice").value;
        data += "&flightDemandSlider="+document.getElementById("flightDemandSlider").value;

        $.get("AirFunctions.Admin.AdminAddFlight", data,
            function(msg){
                temp = JSON.parse(msg);
                if(temp.isAdded == 1){
                    // put something here (?)
                }else{
                    alert("something went wrong");
                }
                closeAddFlightModal();
                canModelMakeTheDistance();
            }
        );

    } else {

    }
}

function updateSliderText(){
    document.getElementById("flightDemandValue").value = document.getElementById("flightDemandSlider").value;
}